mantidproject/mantid
# Limit progress bar updates to 1000Hz maximum

**Description of work.**
Filters out progress massages from algorithms that come in less than 1ms since the last one displayed.  This was really slowing down workbench.

The script attached took over 8s in real time to complete before and now 2-3 sec.
[no_misalignment_m1.zip](https://github.com/mantidproject/mantid/files/4220236/no_misalignment_m1.zip)


**Report to:** Jos Cooper

**To test:**
1. You will need a RELEASE BUILD of workbench
1. Run this script and time it with a watch or external timer (don't trust the elapsed times in mantid)
1. It should have completed in about 3 seconds ish.
1. you should still see some progress bar activity.
``` python
from __future__ import (absolute_import, division, print_function, unicode_literals)
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import time
#Import all the files into mantid, then run this:
import os

def plotandsave(number, filepathtosave=str("d:/data/no_misalignment/Mcstas_detector_")):
    CloneWorkspace(mtd["TOF_PSD_detector_0_0"], OutputWorkspace="DetectorMcstas")
    for i in range(199,-1,-1):
        AppendSpectra("DetectorMcstas","TOF_PSD_detector_0_"+str(i),OutputWorkspace="DetectorMcstas")
    SaveNexus(mtd["DetectorMcstas"],filepathtosave+"\\McStas_detector_"+str(number)+".nxs")

#plotandsave("Offspec_misaligned")


def loadplotandsave(filepath, detector_name, scan_length):
    if scan_length!=1:
        for i in range(scan_length):
            for j in range(200):
                LoadAscii(str(filepath)+"/"+str(i)+"/"+str(detector_name)+str(j)+".dat",OutputWorkspace=str(detector_name)+str(j))
            plotandsave(str(i),filepathtosave=filepath)
            CloneWorkspace("DetectorMcstas",OutputWorkspace="DetectorMcstas"+str(i))
    else:
        for j in range(200):
            print(str(filepath)+"/"+str(detector_name)+str(j)+".dat")
            LoadAscii(str(filepath)+"/"+str(detector_name)+str(j)+".dat",OutputWorkspace=str(detector_name)+str(j))
        plotandsave("0",filepathtosave=filepath)
        CloneWorkspace("DetectorMcstas",OutputWorkspace="DetectorMcstas")


#Using m=1 plus up at far end not down

start_time = time.time()
filepath = r"d:/data/no_misalignment"
loadplotandsave(filepath, "TOF_PSD_detector_0_",1)
RenameWorkspace("DetectorMcstas","Aligned_Mcstas_m1")
elapsed_time=(time.time() - start_time)
print(elapsed_time)
```
5. Run a long running algorithm, "Pause"? and check you still get progress presented

Fixes #27963.
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
